cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,331 @@
+
+    // Arrange all rows of items within the available width and get the total height used.
+    const int totalHeight = layoutAllRows(adjustedRect.x(), adjustedRect.y(), availableWidth);
+
+    // If the layout's parent is a QWidget, update its minimum size to ensure it can accommodate
+    // the arranged items' dimensions.
+    if (QWidget *parentWidgetPtr = parentWidget()) {
+        parentWidgetPtr->setMinimumSize(availableWidth, totalHeight);
+    }
+}
+
+/**
+ * @brief Arranges items in rows based on the available width.
+ *        Items are added to a row until the row's width exceeds `availableWidth`.
+ *        Then, a new row is started.
+ * @param originX The starting x-coordinate for the row layout.
+ * @param originY The starting y-coordinate for the row layout.
+ * @param availableWidth The available width to lay out items.
+ * @return The y-coordinate of the final row's end position.
+ */
+int FlowLayout::layoutAllRows(const int originX, const int originY, const int availableWidth)
+{
+    QVector<QLayoutItem *> rowItems; // Temporary storage for items in the current row.
+    int currentXPosition = originX;  // Tracks the x-coordinate for placing items in the current row.
+    int currentYPosition = originY;  // Tracks the y-coordinate, updated after each row.
+
+    int rowHeight = 0; // Tracks the maximum height of items in the current row.
+
+    // Iterate through all layout items to arrange them.
+    for (QLayoutItem *item : items) {
+        if (item == nullptr || item->isEmpty()) {
+            continue;
+        }
+
+        QSize itemSize = item->sizeHint(); // The suggested size for the item.
+        const int itemWidth = itemSize.width() + horizontalSpacing();
+
+        // Check if the item fits in the current row's remaining width.
+        if (currentXPosition + itemWidth > availableWidth) {
+            // If not, layout the current row and start a new row.
+            layoutSingleRow(rowItems, originX, currentYPosition);
+            rowItems.clear();                                  // Clear the temporary storage for the new row.
+            currentXPosition = originX;                        // Reset x-position to the start of the new row.
+            currentYPosition += rowHeight + verticalSpacing(); // Move y-position down for the new row.
+            rowHeight = 0;                                     // Reset row height for the new row.
+        }
+
+        // Add the item to the current row.
+        rowItems.append(item);
+        rowHeight = qMax(rowHeight, itemSize.height());             // Update the row height to the tallest item.
+        currentXPosition += itemSize.width() + horizontalSpacing(); // Move x-position for the next item.
+    }
+
+    // Layout the final row if there are remaining items.
+    layoutSingleRow(rowItems, originX, currentYPosition);
+
+    // Return the total y-coordinate at the bottom of the last row.
+    return currentYPosition + rowHeight;
```

Comment:
inconsistent format with what's above (`height += rowHeight; // Add the final row's height`)